src/services/helpers.ts

```diff
@@ -9,13 +9,58 @@ export class ValidationError extends Error {
   }
 }
 
+export class TimeoutError extends Error {
+  constructor(message: string) {
+    super(message);
+    this.name = "TimeoutError";
+  }
+}
+
 export async function fetchWithValidator<T, O, I>(
   validator: Type<T, O, I>,
   requestInfo: RequestInfo,
   init?: RequestInit
 ): Promise<T> {
   const response = await fetch(requestInfo, init);
 
+  const json = await response.json();
+
+  if (!response.ok) {
+    throw new Error(json.message ?? "Fetch failed");
+  }
+
+  const decoded = validator.decode(json);
+  return fold(
+    () => {
+      throw new ValidationError(reporter(decoded).join(" "));
+    },
+    (value: T) => value
+  )(decoded);
+}
+
+async function timeout<T>(
+  ms: number,
+  promise: Promise<T>,
+  errorMessage?: string
+): Promise<T> {
+  return new Promise<T>(async (resolve, reject) => {
+    setTimeout(() => {
+      reject(new TimeoutError(errorMessage || "Request timed out"));
+    }, ms);
+    resolve(await promise);
+  });
+}
+
+export async function fetchWithValidatorWithTimeOut<T, O, I>(
```

Comment:
Come to think of it, we don't need an additional function, we can add a `timeout?: boolean | number` param to `fetchWithValidator`, that defaults to `false`. So when we want a fetch to timeout, we can either add `true` (which defaults to `API_CALL_TIMEOUT` or the number of milliseconds.